# Statbotics Python API Documentation

Statbotics.io aims to modernize FRC data analytics through developing and distributing advanced metrics such as Elo and OPR. We have created a Python API to easily query these metrics. This document will serve as a guide to using the API. Enjoy!

The Statbotics API is available on PyPI and can be installed using pip. On your local machine, run **"python -m pip install statbotics"**

Having installed the library, we can now use it in our code. The following cell shows the minimal setup required.

Resources:
Website: https://statbotics.io/
Documentation: https://statbotics.readthedocs.io/en/latest/
Github: https://github.com/statbotics/statbotics
Chief Delphi: https://www.chiefdelphi.com/t/introducing-statbotics-io-modernizing-frc-data-analytics/385360

In [3]:
import statbotics

sb = statbotics.Statbotics()

The Statbotics API lets you query the following objects:
1. **Teams**: Get a team's location and Elo statistics (Current Elo, Recent Elo, Mean Elo, Max Elo)
2. **Years**: Get the prediction accuracy and Brier score using Elo and OPR to predict matches, and ILS to predict RPs
3. **TeamYears**: Get a team's performance in a specific year (Start Elo, Pre-Champs Elo, End Elo, ... , Component OPRs, ILS)
4. **Events**: Get summary statistics for a given event (Type, Week, Top 8 Elos, Top 24 Elos, Top 8 OPRs, Top 24 OPRs)
5. **TeamEvents**: Get a team's performance in a specific event (Start Elo, Pre-Playoffs Elo, End Elo, ..., Component OPRs, ILS)
6. **Matches**: Get statistics for a specific match (score breakdown, Elo, OPR, ILS predictions)
7. **TeamMatches**: Get information for a specific team in a specific match (Match, Alliance, Elo)

For each of these seven objects, there are functions to return a specific object, or a list of objects by certain criteria. Each object is returned as a dictionary mapping columns to specific values. Let's look at some examples!

## getTeam()

The getTeam() function returns information on a specific team. Let's see how we can get the Elo statistics for my team, FRC 5511 Cortechs Robotics.

In [6]:
data  = sb.getTeam(5511)
print(data)
print()

print("Team ", data["team"], "has a current Elo of", data["elo"])

{'team': 5511, 'name': 'Cortechs Robotics', 'state': 'NC', 'country': 'USA', 'district': 'fnc', 'active': True, 'elo': 1702, 'elo_recent': 1658, 'elo_mean': 1603, 'elo_max': 1702}

Team  5511 has a current Elo of 1702


It's that simple! We can see the name, state, district, country, active status, and elo statistics. As a quick primer, Elo is a measure of team strength, updated after each match based on predicted and actual winning margin. An Elo of 1500 is average and an Elo of 1800 corresponds roughly to the 99% percentile.

## getTeams()

The getTeams() function returns information on multiple teams, and supports multiple filters. We can pass arguments to filter by country, state, district, and active status. We can also order the output by a metric, and limit the output to a number of entries to speed up the query. Check https://statbotics.readthedocs.io/en/latest/ for all the details! For now, let's see how we can get the top 10 teams by recent elo.

In [14]:
print("Top 10 teams by Recent Elo")
print()

data = sb.getTeams(metric="elo_recent", limit=10, fields=["team", "elo_recent"])

print("Rank\t Team\t Elo")
for i, entry in enumerate(data):
    print(i+1, "\t", entry["team"], "\t", entry["elo_recent"])

Top 10 teams by Recent Elo

Rank	 Team	 Elo
1 	 254 	 1972
2 	 2056 	 1946
3 	 1678 	 1915
4 	 148 	 1908
5 	 195 	 1886
6 	 2046 	 1876
7 	 118 	 1872
8 	 2481 	 1872
9 	 2910 	 1864
10 	 1114 	 1856


## getYear() and getYears()

These functions return information on year objects. Each dictionary contains the year, accuracy (acc) and mean-squared-error (mse) for match predictions that year (using Elo, OPR, and a mixed approach), and accuracy and mean-squared-error for ranking point predictions (2016-Present) using Caleb Sykes' ILS metric.

In [15]:
print("Years Ranked By Match Prediction Accuracy")
print()

data = sb.getYears(metric="mix_acc", fields=["year", "mix_acc", "mix_mse"])

print("Year\t Accuracy\t MSE")
for entry in data:
    print(entry["year"], "\t", str(entry["mix_acc"]).ljust(6), "\t", entry["mix_mse"])

Years Ranked By Match Prediction Accuracy

Year	 Accuracy	 MSE
2018 	 0.7382 	 0.175
2013 	 0.7332 	 0.1738
2019 	 0.7125 	 0.1791
2020 	 0.7088 	 0.1858
2015 	 0.7075 	 0.1821
2016 	 0.7045 	 0.1841
2011 	 0.7028 	 0.1714
2014 	 0.702  	 0.1889
2012 	 0.6917 	 0.1862
2009 	 0.672  	 0.2004
2006 	 0.6631 	 0.2024
2008 	 0.6605 	 0.1944
2005 	 0.6547 	 0.2089
2017 	 0.6528 	 0.2063
2004 	 0.6316 	 0.2082
2010 	 0.6285 	 0.1774
2003 	 0.6196 	 0.2249
2007 	 0.6041 	 0.2146
2002 	 0.5426 	 0.2442


## getTeamYear() and getTeamYears()

We also can get information for a specific team during a specific year (ex. 254 during 2018). The output dictionary includes the team, year, location information, Elo statistics (Start, Pre-Champs, End, Mean, Max, Diff), and component OPR statistics (using ixOPR with 2 iterations). 

Let's see how we can get the top 25 most dominant seasons of all time (keeping in mind it is easier to dominate in certain years).

In [16]:
print("Top 25 End of Season Elos of All Time")
print()

data = sb.getTeamYears(metric="elo_end", limit=25, fields=["team", "year", "elo_max"])

print("Rank\t Team\t Year\t Elo Max")
for i, entry in enumerate(data):
    print(i + 1, "\t", entry["team"], "\t", entry["year"], "\t", entry["elo_max"])
print()

Top 25 End of Season Elos of All Time

Rank	 Team	 Year	 Elo Max
1 	 469 	 2010 	 2226
2 	 1114 	 2015 	 2181
3 	 1114 	 2008 	 2104
4 	 254 	 2015 	 2145
5 	 1114 	 2010 	 2210
6 	 1114 	 2013 	 2069
7 	 2056 	 2013 	 2061
8 	 254 	 2018 	 2046
9 	 254 	 2014 	 2028
10 	 1678 	 2015 	 2033
11 	 2056 	 2018 	 2022
12 	 2056 	 2015 	 2057
13 	 67 	 2010 	 2037
14 	 330 	 2007 	 2069
15 	 2056 	 2010 	 2014
16 	 25 	 2007 	 2020
17 	 1023 	 2015 	 2015
18 	 987 	 2015 	 1989
19 	 1678 	 2018 	 2001
20 	 25 	 2006 	 1994
21 	 987 	 2013 	 1984
22 	 2056 	 2012 	 1978
23 	 254 	 2010 	 1963
24 	 2056 	 2011 	 2025
25 	 2826 	 2015 	 1963



## getEvent() and getEvents()

These two functions return information on events. For a given event, we can get the key, name, week, type (regional, district, etc), Elo statistics (Top 8, Top 24, Mean), and OPR statistics (Top 8, Top 24, Mean).

This example gets the average "Top 8 OPR" by event type (regional, district, district champ, worlds division, einstein)

In [20]:
print("Average Top 8 Elo by Event Type, 2019")
print()

types = {
    0: "regional",
    1: "district",
    2: "district champs",
    3: "worlds division",
    4: "einstein",
}

print("Type\t\t\t Top 8 OPR")
for key, name in types.items():
    data = sb.getEvents(year=2019, type=key, fields=["opr_top8"])
    average = round(sum([r["opr_top8"] for r in data]) / len(data), 2)
    print(name.ljust(15), "\t", average)

Average Top 8 Elo by Event Type, 2019

Type			 Top 8 OPR
regional        	 23.22
district        	 21.52
district champs 	 27.56
worlds division 	 33.2
einstein        	 31.83


## getTeamEvent() and getTeamEvents()

Similar to getTeamYears(), we can get statistics on how a team fared in a specific event. This information includes the team, event information (week, type, etc), Elo statistics (Start, Pre-Playoffs, End, Mean, Max, Diff), and component OPR statistics.

Let's take a look at 469's start and end Elo per event during the 2010 season, which remains the highest Elo ratings ever recorded in FRC.

In [23]:
print("469 Elo through 2010 Events")
print()

data = sb.getTeamEvents(team=469, year=2010, metric="-time", fields=["event", "elo_start", "elo_end"])

print("Event Key\t Start\t End Elo")
for entry in data:
    print(entry["event"].ljust(10), "\t", entry["elo_start"], "\t", entry["elo_end"])
print()

469 Elo through 2010 Events

Event Key	 Start	 End Elo
2010dt1    	 1668 	 1758
2010oc     	 1758 	 2132
2010gl     	 2132 	 2185
2010cur    	 2185 	 2148
2010cmp    	 2148 	 2131



## getMatch() and getMatches()

If you are interested in score breakdowns and individual predictions, these are the functions for you! These methods return a lot of information on the match (elims, set number, match number), scores (component breakdown for both alliances), and predictions (RP predictions and predicted winner by Elo, OPR, mixed method)

We can see how Statbotics.io predicted matches for 2019 Houston Einstein Championships (Spoiler Alert: 78% Accuracy)

In [39]:
print("Match Predictions for 2019 Houston Champs")
print()

matches = sb.getMatches(event="2019cmptx", elims=True, fields=["key", "winner", "mix_winner", "mix_win_prob"])

print("Match Key\t\t Actual\t\t Predicted\tRed Win Prob")
for match in matches:
    print(match["key"].ljust(15), "\t", match["winner"], "\t\t", match["mix_winner"], "\t\t", match["mix_win_prob"])

Match Predictions for 2019 Houston Champs

Match Key		 Actual		 Predicted	Red Win Prob
2019cmptx_sf1m1 	 blue 		 blue 		 0.31
2019cmptx_sf1m2 	 red 		 red 		 0.76
2019cmptx_sf1m3 	 blue 		 blue 		 0.16
2019cmptx_sf1m4 	 blue 		 red 		 0.58
2019cmptx_sf1m5 	 red 		 blue 		 0.41
2019cmptx_sf1m6 	 red 		 red 		 0.62
2019cmptx_sf1m7 	 blue 		 blue 		 0.22
2019cmptx_sf1m8 	 blue 		 blue 		 0.35
2019cmptx_sf1m9 	 red 		 red 		 0.6
2019cmptx_sf1m10 	 blue 		 blue 		 0.43
2019cmptx_sf1m11 	 red 		 red 		 0.71
2019cmptx_sf1m12 	 blue 		 blue 		 0.26
2019cmptx_sf1m13 	 blue 		 red 		 0.53
2019cmptx_sf1m14 	 blue 		 blue 		 0.36
2019cmptx_sf1m15 	 red 		 red 		 0.87
2019cmptx_f1m1  	 red 		 blue 		 0.37
2019cmptx_f1m2  	 blue 		 blue 		 0.39
2019cmptx_f1m3  	 blue 		 blue 		 0.37


## getTeamMatch() and getTeamMatches()

Our final two methods, these are mostly here for completion, though I'm sure there are creative uses for these. Each dictionary only returns the team, match, alliance, and Elo prior to the match being played. 

In this example, we see how many matches Team 125 plays per year (they've averaged 100+ since 2014!)

In [41]:
print("Team 125 Matches played by Year")
print()

data = sb.getTeamMatches(team=125, fields=["year"])

years = {i: 0 for i in range(2002, 2020)}
for match in data:
    years[match["year"]] += 1
    
print("Year\t Matches")
for year, matches in years.items():
    print(year, "\t", matches)
print()

Team 125 Matches played by Year

Year	 Matches
2002 	 18
2003 	 27
2004 	 20
2005 	 8
2006 	 29
2007 	 24
2008 	 26
2009 	 20
2010 	 19
2011 	 34
2012 	 41
2013 	 62
2014 	 98
2015 	 99
2016 	 107
2017 	 107
2018 	 97
2019 	 103



That's all for this demonstration. I hope you have a better understanding of how the API works, and I look forward to seeing all the cool things everyone does with this!